### ISSUE TITLE
Consider converting `*Subtensor*` indices to unsigned integers

### ISSUE THREAD

**Issue by brandonwillard** [aesara-devs/aesara#1147]
We could apply an unsigned int dtype conversion to the index inputs of `*Subtensor*` index arguments in an attempt to get some small performance improvements&mdash;as mentioned in https://github.com/aesara-devs/aesara/pull/1081#issuecomment-1206589448.  

One bit of reasoning for this is that some architectures might produce faster code by removing unnecessary instructions when an unsigned integer is used for indexing (e.g. omitting a ["sign extension"](https://stackoverflow.com/questions/55604029/should-i-always-use-size-t-when-indexing-arrays/55604267#55604267)).

Since a dtype conversion at run-time could be costly, we might need to restrict such a rewrite to only constant index inputs.  Anyone implementing this might want to first check that, though.

Also, this dtype conversion isn't universally helpful.  On my machine, only Numba sees a small benefit from it, so we would need to determine whether or not it's even worth implementing in Aesara, instead of as [a rewrite in Numba itself](https://numba.readthedocs.io/en/stable/developer/rewrites.html) (i.e. not in our transpiled Numba code).  Furthermore, we should first check with the Numba community and determine why this isn't already an optimization on their (or LLVM's) end.

In the meantime, if one wants this potential improvement, they can manually set their index variables to unsigned integers&mdash;especially when/if they're constants.

**Comment by aseyboldt** [comment]
I don't think the performance improvement is due to sign extension, but it is because of the special behavior of negative indices in numpy ie that `array[-1]` refers to the last element. This means that llvm can not reach the correct address by just computing `array.ctypes.data + index`, but has to do some computation involving `index` and `len(array)` to find the correct offset.

By using an unsigned integer we implicitly promise that the index won't be negative, so that computation isn't necessary anymore.

This should only make sense if the index array is constant and non-negative, so we only want it in those cases: If the index isn't constant, how would we know if it contains negative indices? We'd have to iterate through it, and I'd expect that would be slower than the performance improvement we can get by avoiding the extra computation in the loop.

Also, the same does happen for jax for me:

```python
import jax.numpy as jnp
import numpy as np

idx = np.random.randint(100, size=10_000)
idx1 = idx.astype("i2")  # signed integer
idx2 = idx.astype("u2")  # unsigned integer
vals = jnp.arange(100, dtype="float32")

%timeit vals[idx]
819 µs ± 57 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit vals[idx1]
862 µs ± 16.2 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit vals[idx2]
597 µs ± 14.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit vals[idx3]
593 µs ± 24.8 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Even though I somehow don't see the benefit for smaller dtypes I saw in numba. (we can find the correct type using `np.min_scalar_type` as in the old PR). By the way, if someone still wants to use that code, that's fine by me.

**Comment by brandonwillard** [comment]
> I don't think the performance improvement is due to sign extension, but it is because of the special behavior of negative indices in numpy ie that `array[-1]` refers to the last element.

Yes, omitting the "wrap-around" logic will likely be the main improvement for _Python-based indexing_, but that sign extension is all I can currently think of for the more general case of indexing (e.g. array indexing in C/C++ per the C backend).   Since the Numba backend was already covered in the comment to which I linked, the C backend is the next thing that could use some speculation regarding why this might be an improvement.

The `*Subtensor*` C implementations can be a little too indirect to tell from a glance, so it's not entirely clear to me whether or not wrap-around can be avoided by only a dtype change, or if something much more general&mdash;like that sign extension example&mdash;might be the only kind of thing that could apply to the existing implementation.  In other words, changes to the C implementations might be needed in order to see improvements under this optimization.

As I mentioned, a quick inspection didn't reveal an improvement for the C backend, so that could be the case, or it could be my architecture.  No idea.  Regardless, we need someone to take a closer look at the C backend situation in order to justify a broad graph-level rewrite.


> Also, the same does happen for jax for me:

Thanks for reporting that.  It looks like we should consider this for the JAX case as well.

We first need to figure out exactly _how_ we want to implement this (e.g. as a rewrite that's only applied in Numba mode, as a tweak to our Numba transpilation code/settings, as a general rewrite that applies to all backends, etc.)  For example, if this optimization also applies to JAX, then we might want to lean toward mode-specific rewrites, and not two separate implementations at the transpilation level.

> Even though I somehow don't see the benefit for smaller dtypes I saw in numba. (we can find the correct type using `np.min_scalar_type` as in the old PR).

I don't see why that couldn't be done for constants anyway.

**Comment by aseyboldt** [comment]
>> Even though I somehow don't see the benefit for smaller dtypes I saw in numba. (we can find the correct type using np.min_scalar_type as in the old PR).

> I don't see why that couldn't be done for constants anyway.

A general rewrite that changes the dtype of all constant tensors could change results, unless we are very careful in all ops that use them. Let's look at this graph:

```
a = at.as_tensor_variable(127, dtype="int32")
b = at.as_tensor_variable(1, dtype="int32")
out = a + b
```
This should return 128, but if we change each tensor constant to the smallest possible dtype, and naively run the addition in that dtype we'd get -1 because of wraparound. It might be possible to avoid this problem, eg by storing the original dtype of the values and casting to that dtype when we use the value later, for instance in an elemwise, but I think this might turn out to be a bit more trouble than it is worth.

Which also reminds me of this nice code in the current aesara (used to be the same in theano):

![image](https://user-images.githubusercontent.com/1882397/186794620-f72ed749-97c8-4ab6-ad26-d5939dd66126.png)

This sure has some logic to it, but I think we might be better off and avoid potential bugs by defaulting to int64 for integer types...

> Yes, omitting the "wrap-around" logic will likely be the main improvement for Python-based indexing, but that sign extension is all I can currently think of for the more general case of indexing (e.g. array indexing in C/C++ per the C backend). Since the Numba backend was already covered in the comment to which I linked, the C backend is the next thing that could use some speculation regarding why this might be an improvement.

Well, if the C backend handles negative indices correctly (and I think it does) then that logic has to be somewhere, and it also should come at the same cost as in numba. I think the C implementation at least of `AdvancedIncSubtensor1` (but also AdvancedSubtensor` I think) are pretty slow for other reasons, so even if that performance benefit is there, it might be hidden in the noise of some other overheads.

**Comment by ricardoV94** [comment]
> Furthermore, we should first check with the Numba community and determine why this isn't already an optimization on their (or LLVM's) end.

Yeah, I am curious why that isn't done. Seems like an easy one.

**Comment by brandonwillard** [comment]
> A general rewrite that changes the dtype of all constant tensors could change results, unless we are very careful in all ops that use them.

We're talking about constants that are `*Subtensor*` indices, so, if they can't be safely changed, then we shouldn't be dealing with them here.

> Which also reminds me of this nice code in the current aesara (used to be the same in theano):
> 
> ![image](https://user-images.githubusercontent.com/1882397/186794620-f72ed749-97c8-4ab6-ad26-d5939dd66126.png)
> 
> This sure has some logic to it, but I think we might be better off and avoid potential bugs by defaulting to int64 for integer types...
> 
> >

Yes, that's a bad one.  We need to finish implementing Theano's NumPy casting policy, and use it.


> Well, if the C backend handles negative indices correctly (and I think it does) then that logic has to be somewhere

The implementations do; I just haven't looked closely enough to see if it has or uses logic that could avoid those checks and/or index normalizations based on the indices' dtypes.

**Comment by aseyboldt** [comment]
> Yeah, I am intrigued by this one. Seems like an easy optimization.

I looked a little bit into the constant propagation llvm will do:

<details>

```python
import numpy as np
import numba


# Check is optimized away
consts = np.zeros(10)
@numba.njit
def test_fn():
    if consts[1] == 0:
        return 0
    else:
        return 1


# Loop is not optimized away
consts = np.ones(1000)
@numba.njit
def test_fn():
    is_zero = True
    for idx in range(len(consts)):
        if consts[idx] != 0:
            is_zero = False
    return is_zero


# Loop is optimized away
consts = np.ones(1000, dtype=bool)
@numba.njit
def test_fn():
    is_zero = True
    for idx in range(len(consts)):
        if consts[idx]:
            is_zero = False
    return is_zero


# Condition is not optimized away
consts = np.zeros(1000, dtype=int)
@numba.njit
def test_fn2(out):
    for i in range(len(consts)):
        if consts[i] > 0:
            out[i] = 1
        else:
            out[i] = -1
```

</details>

So it clearly does *some* optimizations, but it doesn't seem to be smart enough to notice the properties it would need for this, and I don't think this optimization actually is *that* easy from the point of view of llvm:

We can have a look at the optimized llvm code for the last function, which I think is close enough to the indexing we are doing (get the complete llvm code with `test_fn.inspect_llvm()[test_fn.signatures[0]]` after running it once, and while disabling simd to not make it needlessly complicated, so `%env NUMBA_LOOP_VECTORIZE=0`):

```
; There is a constant data section, where it just stores the bytes of our constant array
@.const.array.data.1 = internal unnamed_addr constant [8000 x i8] c"\01\00\00\00\00\00\00\00\01...."

; In the loop of the function (simplified a bit to get rid of the partial loop unrolling):

  ; %4 is the loop index

  ; Compute the address of the current const data item, ie %scevgep17 = const.ctypes.data + idx * const.dtype.itemsize
  %scevgep17 = getelementptr [8000 x i8], [8000 x i8]* @.const.array.data.1, i64 0, i64 %4

  ; Cast the pointer from a byte pointer to an i64 integer pointer
  %scevgep1718 = bitcast i8* %scevgep17 to i64*

  ; Get the content of the constant array at that location
  %.233.4 = load i64, i64* %scevgep1718, align 8

  ; Compare it to 0
  %.237.4 = icmp sgt i64 %.233.4, 0

  ; Get the pointer to the output array at the current location
  %scevgep = getelementptr double, double* %arg.out.4, i64 %4

  ; Choose 1 or -1 based on the previous comparison
  %..4 = select i1 %.237.4, double 1.000000e+00, double -1.000000e+00

  ; Write the value to the output array location
  store double %..4, double* %scevgep, align 8
```

We now like llvm to optimize away the comparison. In this particular example it would be enough if it noticed that all the `load` operations return the same value, because for all valid values of `%4` the values in the constant array are the exact same. But notice the "for all valid values of %4" here. I suspect that llvm does not check this, and I can't say I'm too surprised.
In our original example this wouldn't even be enough. There, it would have to notice that while the `load` instruction can yield different values, for all of *those* values the next instruction (`icmp`) returns zero. I don't think this counts as an "easy optimization" anymore. :-)

One thing I would have expected llvm to optimize is if `consts = np.zeros(1000)`, because in that case the constant section becomes

```
@.const.array.data.1 = internal unnamed_addr constant [8000 x i8] zeroinitializer, align 8
```

So there it wouldn't be too hard to see that this is always zero, I think?

```
  %scevgep17 = getelementptr [8000 x i8], [8000 x i8]* @.const.array.data.1, i64 0, i64 %4
  %scevgep1718 = bitcast i8* %scevgep17 to i64*
  %.233.4 = load i64, i64* %scevgep1718, align 8
```

This still doesn't happen even if we get rid of the bitcast:

```
  %scevgep17 = getelementptr [1000 x i8], [1000 x i8]* @.const.array.data.1, i64 0, i64 %lsr.iv
  %.233.4 = load i8, i8* %scevgep17, align 1
```

So either llvm is missing something relatively simple (?) in this simplified example, or there is some piece to the puzzle I just don't know.
I'm really getting to the limit of what I know about llvm though, I think if we want to investigate this further we need someone with more detailed knowledge about the compiler optimization passes than I have.

But for the most part, it doesn't feel like llvm is the place where an optimization like this actually is easy, it feels to me like aesara really is the correct location for an optimization like this.

**Comment by brandonwillard** [comment]
> But for the most part, it doesn't feel like llvm is the place where an optimization like this actually is easy, it feels to me like aesara really is the correct location for an optimization like this.

LLVM might not be, but Numba possibly.  I put a link in the opening comment to Numba's rewrite framework.

**Comment by brandonwillard** [comment]
We should consider changing the dtypes of `Alloc`, `ARange`, and similar allocation `Op`s that are used as indices&mdash;at least when they have no other clients.

**Comment by brandonwillard** [comment]
One thing we could also explore (separately) is converting constant negative indexes to positive indexes when the shape of the tensor is fixed. We know this at the Aesara level, but not necessarily after compilation.

_Originally posted by @ricardoV94 in https://github.com/aesara-devs/aesara/issues/1150#issuecomment-1229222356_
